lab_3/main.py

```diff
@@ -20,7 +21,44 @@ def tokenize_by_sentence(text: str) -> tuple:
          (('_', 'h', 'e', '_'), ('_', 'i', 's', '_'), ('_', 'h', 'a', 'p', 'p', 'y', '_'))
          )
     """
-    pass
+    if not isinstance(text, str) or not text:
+        return ()
+
+    # split text by sentence
+    text = text.lower()
+    text = re.split(r'[.!?] ', text)
+
+    # tokenization of sentence
+    deutsch_letters = {'ö': 'oe', 'ü': 'ue', 'ä': 'ae', 'ß': 'ss'}
+    clean_text = []
+    clean_sentence = ''
+    for sentence in text:
+        for symbol in sentence:
+            if symbol in deutsch_letters:  # replace umlauts and ß
+                clean_sentence += deutsch_letters[symbol]
+            elif symbol.isalpha() or symbol.isspace():
+                clean_sentence += symbol
+        clean_text.append(clean_sentence.split())
+        clean_sentence = ''
+
+    split_tokens = []  # list for sentence consisted of split tokens
+    token = []  # list for split word
```

Comment:
You can use the local `for` loop variables instead of creating other variables outside the loop.

Reply:
Done